src/main/Scheduler.java

```diff
@@ -64,13 +74,52 @@ public Scheduler(int floorEventPort, int arrPort, int destPort, int floorPort, i
 	@Override
 	public void run() {
 		while (true) {
+			getTimeoutEvent();
 			currentState.handleFloorEvent();
 			currentState.handleDestinationEvent();
 			currentState.handleArrivalEvent();
 		}
 	}
-
 	
+	/**
+	 * Start a timer for elevator
+	 */
+	public void startElevatorTimer(int elevatorId, boolean beforeArrivedAtSrcFlr, int numFloors) {
+		timers[elevatorId] = new SchedulerTimer(this, Integer.toString(elevatorId), beforeArrivedAtSrcFlr, elevatorId, timerPort);
+		timers[elevatorId].start(numFloors);
+	}
+	
+	/**
+	 * @return the timers
+	 */
+	public void cancelElevatorTimer(int elevatorId) {
+		timers[elevatorId].cancel();
+	}
+	
+	public void permanentFault(TimeoutEvent t) {
+		ArrayList<FloorEvent> floorEventRemove = new ArrayList<FloorEvent>();
+		ArrayList<Event> destEventRemove = new ArrayList<Event>();
+
+		for(FloorEvent e: sentFloorEvents) {
+			if(e.getElevatorId() == t.getElevatorId()) {
+				e.setErrorCode(0);
+				floorEventRemove.add(e);
+			}
+		}
+		floorEvents.addAll(floorEventRemove);
+		sentFloorEvents.removeAll(floorEventRemove);

```

Comment:
Can you remove events in the for loop instead of doing this?

Reply:
nope, sadly it causes a really ugly error